In [2]:
import pandas as pd
import json

df = pd.read_parquet("transactions.parquet")
df['year'] = df['Order_Date_FK'].astype(str).str[:4]
df['month'] = df['Order_Date_FK'].astype(str).str[4:6]
df['day'] = df['Order_Date_FK'].astype(str).str[6:]

# values = dict()

# for col in df.columns:
#     values[col] = list(df[col].astype(str).unique())

# values

In [5]:
with open("values.json", "w") as f:
    json.dump(values, f)

In [6]:
import sqlite3 as sq
import pandas as pd

connection = sq.connect("database.db")

In [3]:
df

Shop  Order_Date_FK          ProductCode  OriginalSaleAmountInclVAT  \
0          36       20210730 -1968361059464632550                      99.95   
1          58       20210730 -1745497788861904687                      21.95   
2          49       20210730 -4482110002029860032                      69.95   
3          49       20210730 -9146072753857637799                      99.95   
4          15       20210730 -8829485741699345914                      49.95   
...       ...            ...                  ...                        ...   
1759884    79       20210730  -745373077535329727                     120.00   
1759885    79       20210730 -2694142533085979238                       8.50   
1759886    79       20210730  -745373077535329727                     120.00   
1759887    60       20210730  1659237309319929690                      65.95   
1759888    48       20210730  4579744411951681005                      69.95   

                  CustomerID  SaleDocumentNumber  Returned  RevenueInclVAT  \
0       -2190786785520839526            23995792         0           74.96   
1       -3499949798555834600            23995794         0           21.95   
2       -7841417233849937172            23995797         0           38.47   
3       -7841417233849937172            23995797         0           54.97   
4        7033964597104594686            23995798         0           17.48   
...                      ...                 ...       ...             ...   
1759884  -299104793296959079            23995789         0           78.00   
1759885  -299104793296959079            23995789         0            8.50   
1759886  -299104793296959079            23995789         0           78.00   
1759887 -4340188327539121409            23995790         0           42.87   
1759888  6389575287796397740            23995791         0           69.95   

         CostPriceExclVAT            BrandName           ModelGroup  \
0               36.534515  3694837121284491212  3162564956579801398   
1                7.364523 -1657270726128381662 -6715310361529623819   
2               24.415000 -7202472198306336051  -248459121805647149   
3               32.293506 -2074012670196538378  -248459121805647149   
4                8.885632 -4869459690934553786 -4124706989267335995   
...                   ...                  ...                  ...   
1759884         38.390231  4614247681931237507 -4124706989267335995   
1759885          2.328775 -8892132782483441489 -2426224035638420213   
1759886         38.390231  4614247681931237507 -4124706989267335995   
1759887         22.288649 -4445228402666777776  3162564956579801398   
1759888         24.437763 -3001169043836376081 -4124706989267335995   

                ProductGroup  year month day  
0        -453682476182549203  2021    07  30  
1        -453682476182549203  2021    07  30  
2        5294166889150293396  2021    07  30  
3        5294166889150293396  2021    07  30  
4        -453682476182549203  2021    07  30  
...                      ...   ...   ...  ..  
1759884  -453682476182549203  2021    07  30  
1759885  5965103661511607192  2021    07  30  
1759886  -453682476182549203  2021    07  30  
1759887  -453682476182549203  2021    07  30  
1759888  -453682476182549203  2021    07  30  

[1759889 rows x 15 columns]

#### Duplicates are unclear. It would be better to know why there are duplicates, and be certain how to handle them correctly.

In [7]:
duplicates = pd.read_sql_query(
"""
SELECT *
FROM trans
WHERE SaleDocumentNumber = 23719501
""", connection
)
duplicates

index  Shop  Order_Date_FK          ProductCode  \
0    1005963    49       20210701 -5223360592088065790   
1    1005964    49       20210701  3510254952241336477   
2    1005965    49       20210701   449823314919264450   
3    1005966    49       20210701  3274348471310059289   
4    1005967    49       20210701 -5584415452890875547   
..       ...   ...            ...                  ...   
97   1349146    49       20210701 -2854091718625257153   
98   1349147    49       20210701 -8041448267277774414   
99   1349148    49       20210701 -7462286497937572848   
100  1349149    49       20210701   441949993592162707   
101  1349150    49       20210701   343948556986501671   

     OriginalSaleAmountInclVAT           CustomerID  SaleDocumentNumber  \
0                       105.00 -1294657933607279032            23719501   
1                        75.95 -1294657933607279032            23719501   
2                        59.95 -1294657933607279032            23719501   
3                       170.00 -1294657933607279032            23719501   
4                       130.00 -1294657933607279032            23719501   
..                         ...                  ...                 ...   
97                       65.95 -1294657933607279032            23719501   
98                       75.95 -1294657933607279032            23719501   
99                      175.00 -1294657933607279032            23719501   
100                     130.00 -1294657933607279032            23719501   
101                      95.95 -1294657933607279032            23719501   

     Returned  RevenueInclVAT  CostPriceExclVAT            BrandName  \
0           0           10.50         25.608496  1168115093187640245   
1           0            7.60         24.500000    33538854880025124   
2           0            6.00         18.950000  9178759879362980169   
3           0           17.00         62.036017  -424472638565217610   
4           0           13.00         36.000000 -7177370704634981598   
..        ...             ...               ...                  ...   
97          0            6.59         25.714611  2776326510072975976   
98          0            7.59         17.493915  1168115093187640245   
99          0           17.50         64.152000 -1873787772202565890   
100         0           13.00         46.889754  1857918515453071333   
101         0            9.59         38.004118  8855638714846779199   

              ModelGroup        ProductGroup  
0   -4124706989267335995 -453682476182549203  
1   -4124706989267335995 -453682476182549203  
2   -4124706989267335995 -453682476182549203  
3   -4124706989267335995 -453682476182549203  
4   -4124706989267335995 -453682476182549203  
..                   ...                 ...  
97  -4124706989267335995 -453682476182549203  
98  -4124706989267335995 -453682476182549203  
99  -4124706989267335995 -453682476182549203  
100 -4124706989267335995 -453682476182549203  
101  3162564956579801398 -453682476182549203  

[102 rows x 13 columns]

In [8]:
duplicates = pd.read_sql_query(
"""
SELECT
    COUNT(ProductCode) AS uniques
FROM trans
WHERE SaleDocumentNumber = 23719501
GROUP BY ProductCode
ORDER BY uniques DESC
""", connection
)
duplicates.head(10)

uniques
0        4
1        4
2        4
3        4
4        2
5        2
6        2
7        2
8        2
9        2

#### Now let's take a look at returns in various ways.

In [9]:
df["Returned"].sum()

67795

In [10]:
returns = pd.read_sql_query(
"""
WITH a AS (
SELECT
    COUNT(Returned) AS C,
    SUM(Returned) AS S,
    ProductCode, BrandName,
    OriginalSaleAmountInclVAT
FROM trans
GROUP BY ProductCode
ORDER BY C DESC
)
SELECT
    CAST(S AS FLOAT) / C AS PercentReturned,
    C AS TotalItems,
    S AS TotalReturns,
    ProductCode, BrandName, OriginalSaleAmountInclVAT
FROM a
ORDER BY TotalReturns DESC, PercentReturned DESC
""", connection)

print("\tGrouped by PoductCode, PercentReturned > 0.2, ordered by TotalReturns descending")
returns[returns["PercentReturned"] > 0.2][:10]

	Grouped by PoductCode, PercentReturned > 0.2, ordered by TotalReturns descending


PercentReturned  TotalItems  TotalReturns          ProductCode  \
0           0.326376        1054           344 -3711611854324721442   
11          0.268102         511           137  4754863614336515610   
14          0.210708         579           122  5587091866033151813   
28          0.321070         299            96  4101533479629837146   
37          0.231405         363            84 -3103717362647513544   
38          0.432292         192            83  4664267319375427392   
58          0.250923         271            68  4510119518088610237   
98          0.412214         131            54  8706358678450957216   
110         0.481481         108            52  1688912043498932481   
159         0.200913         219            44 -8680557187301822340   

               BrandName  OriginalSaleAmountInclVAT  
0    6164254645951948400                      45.95  
11  -1671135878886367730                      25.95  
14   3147597281999378922                      15.95  
28  -3734237314099258167                      75.95  
37  -8710939345520521313                      39.95  
38  -8846961206607438497                      75.95  
58   6164254645951948400                      69.95  
98  -1671135878886367730                      27.95  
110 -8507421977918128293                      95.95  
159 -1873787772202565890                      75.95

#### It is quite possible that product with ProductCode -3711611854324721442 is bad.

In [11]:
returns = pd.read_sql_query(
"""
WITH a AS (
SELECT
    COUNT(Returned) AS C,
    SUM(Returned) AS S,
    ModelGroup
FROM trans
GROUP BY ModelGroup
ORDER BY C DESC
)
SELECT
    CAST(S AS FLOAT) / C AS PercentReturned,
    C AS TotalItems,
    S AS TotalReturns,
    ModelGroup
FROM a
ORDER BY TotalReturns DESC, PercentReturned DESC
""", connection)

print("\tGrouped by ModelGroup, PercentReturned > 0.2, ordered by TotalReturns descending")
returns[:10]

	Grouped by ModelGroup, PercentReturned > 0.2, ordered by TotalReturns descending


PercentReturned  TotalItems  TotalReturns           ModelGroup
0         0.054149      564852         30586  3162564956579801398
1         0.037423      703042         26310 -4124706989267335995
2         0.044297       64497          2857 -6715310361529623819
3         0.039762       38580          1534 -1330987452709117740
4         0.020803       71818          1494  -248459121805647149
5         0.036640       34771          1274  8676381074854750838
6         0.060680       20188          1225 -1172740817228586022
7         0.035239       14728           519  -965590225519141271
8         0.015010       21986           330 -2375149457939821456
9         0.037417        8285           310 -3170699624988567198

In [12]:
returns = pd.read_sql_query(
"""
WITH a AS (
SELECT
    COUNT(Returned) AS C,
    SUM(Returned) AS S,
    BrandName
FROM trans
GROUP BY BrandName
ORDER BY C DESC
)
SELECT
    CAST(S AS FLOAT) / C AS PercentReturned,
    C AS TotalItems,
    S AS TotalReturns,
    BrandName
FROM a
ORDER BY TotalReturns DESC, PercentReturned DESC
""", connection)

print("\tGrouped by BrandName, PercentReturned > 0.2, ordered by TotalReturns descending")
returns[returns["PercentReturned"] > 0.2][:10]

	Grouped by BrandName, PercentReturned > 0.2, ordered by TotalReturns descending


PercentReturned  TotalItems  TotalReturns            BrandName
53          0.251429        1050           264 -8507421977918128293
98          0.249417         429           107   611049624476511346
136         0.221122         303            67  6862799060630055392
139         0.278481         237            66  8690259017905730373
140         0.216393         305            66  7788141470214192600
165         0.343066         137            47 -7130095634701368819
166         0.218605         215            47 -4606403254425245430
184         0.234568         162            38 -7921480381513708220
195         0.430556          72            31  4717096102405955547
204         0.333333          78            26 -7827463481178904503

#### Notice how BrandName -8507421977918128293 has 25% returns on 1000 items. That seems like a bad brand to me. Particularly their product 1688912043498932481 with 50% returns on around 100 items. Maybe it's worth looking into.

In [13]:
results = pd.read_sql_query(
"""
WITH a AS (
SELECT
    COUNT(Returned) AS C,
    SUM(Returned) AS S,
    ProductCode
FROM trans
WHERE BrandName = -8507421977918128293
GROUP BY ProductCode
ORDER BY C DESC
)
SELECT
    CAST(S AS FLOAT) / C AS PercentReturned,
    C AS TotalItems,
    S AS TotalReturns,
    ProductCode
FROM a
ORDER BY TotalReturns DESC, PercentReturned DESC
""", connection)

print("\tGrouped by PoductCode, PercentReturned > 0.2, ordered by TotalReturns descending")
results[results["PercentReturned"] > 0.][:10]

	Grouped by PoductCode, PercentReturned > 0.2, ordered by TotalReturns descending


PercentReturned  TotalItems  TotalReturns          ProductCode
0         0.481481         108            52  1688912043498932481
1         0.306452          62            19 -2589790336782400676
2         0.410256          39            16  -575353219603877387
3         0.238095          63            15 -4297715029242917879
4         0.279070          43            12  2104612900236091154
5         0.184615          65            12  8667661596481621898
6         0.277778          36            10  7884848769945646548
7         0.263158          38            10 -3515954374243214119
8         0.346154          26             9 -1107135324166745045
9         0.166667          54             9 -5774003379652257129

# So what about shops?

In [31]:
returns_per_shop = pd.read_sql_query(
"""
WITH a AS (
SELECT
    COUNT(Returned) AS C,
    SUM(Returned) AS S,
    Shop
FROM trans
GROUP BY Shop
ORDER BY C DESC
)
SELECT
    CAST(S AS FLOAT) / C AS PercentReturned,
    C AS TotalItems,
    S AS TotalReturns,
    Shop
FROM a
ORDER BY PercentReturned DESC
""", connection)

print("\tGrouped by Shop, ordered by PercentReturned descending")
returns_per_shop[:10]

	Grouped by Shop, ordered by PercentReturned descending


PercentReturned  TotalItems  TotalReturns  Shop
0         0.186977      140456         26262   102
1         0.164588        8901          1465   103
2         0.064000         125             8    41
3         0.039513       21512           850    30
4         0.037832       22838           864    68
5         0.035541       11930           424    33
6         0.033156       18820           624     6
7         0.031596       14812           468    43
8         0.031090       18720           582    47
9         0.029993       23072           692    70

# Any customers of interest?

In [15]:
returns_per_customer = pd.read_sql_query(
"""
WITH a AS (
SELECT
    COUNT(Returned) AS C,
    SUM(Returned) AS S,
    CustomerID
FROM trans
GROUP BY CustomerID
ORDER BY C DESC
)
SELECT
    CAST(S AS FLOAT) / C AS PercentReturned,
    C AS TotalItems,
    S AS TotalReturns,
    CustomerID,
    S / 67795.0 AS PercentOfAllReturns
FROM a
ORDER BY TotalReturns DESC, PercentReturned DESC
""", connection)

print("\tGrouped by CustomerID, ordered by TotalReturns and PercentReturned descending")
returns_per_customer

	Grouped by CustomerID, ordered by TotalReturns and PercentReturned descending


PercentReturned  TotalItems  TotalReturns           CustomerID  \
0              0.143486       54277          7788 -5287955982418449235   
1              0.129367        3950           511  1727613839816903509   
2              0.029466        7602           224  4572026730515361595   
3              0.022233        7961           177 -4279123115125507326   
4              0.021077        7022           148  6455628090489614628   
...                 ...         ...           ...                  ...   
347857         0.000000           1             0 -9217944974532867625   
347858         0.000000           1             0 -9219368559219805324   
347859         0.000000           1             0 -9219371575752114423   
347860         0.000000           1             0 -9221298982193402923   
347861         0.000000           1             0 -9221850624083161356   

        PercentOfAllReturns  
0                  0.114876  
1                  0.007537  
2                  0.003304  
3                  0.002611  
4                  0.002183  
...                     ...  
347857             0.000000  
347858             0.000000  
347859             0.000000  
347860             0.000000  
347861             0.000000  

[347862 rows x 5 columns]

#### Customer -5287955982418449235 is pretty huge, and responsible for 11% of all returns.